x-pack/plugins/endpoint/server/routes/alerts/list/lib.ts

```diff
@@ -0,0 +1,233 @@
+    return queries[0];
+  }
+
+  return {
+    match_all: {},
+  };
+}
+
+export const buildAlertListESQuery = async (reqData: AlertListRequestData): Promise<JsonObject> => {
+  const DEFAULT_TOTAL_HITS = 10000;
+  let totalHitsMin: number = DEFAULT_TOTAL_HITS;
+
+  // Calculate minimum total hits set to indicate there's a next page
+  // TODO: handle this for search_after?
+  if (reqData.fromIndex !== undefined) {
+    totalHitsMin = Math.max(reqData.fromIndex + reqData.pageSize * 2, DEFAULT_TOTAL_HITS);
+  }
+
+  const reqBody: AlertListRequestBody = {
+    track_total_hits: totalHitsMin,
+    query: buildQuery(reqData),
+    sort: [
+      // User-defined primary sort, with default to `@timestamp`
+      {
+        [reqData.sort]: {
+          order: reqData.order as 'asc' | 'desc',
+        },
+      },
+      // Secondary sort for tie-breaking
+      {
+        'event.id': {
+          order: reqData.order as 'asc' | 'desc',
+        },
+      },
+    ],
+  };
+
+  const reqWrapper: AlertListRequest = {
+    size: reqData.pageSize,
+    index: EndpointAppConstants.ALERT_INDEX_NAME,
+    body: reqBody,
+  };
+
+  if (reqData.fromIndex !== undefined) {
+    reqWrapper.from = reqData.fromIndex;
+  }
+
+  if (reqData.searchAfter !== undefined) {
+    reqWrapper.body.search_after = reqData.searchAfter;
+  }
+
+  if (reqData.searchBefore !== undefined) {
+    // Reverse sort order for search_before functionality
+    const order: string = reqWrapper.body.sort[0][reqData.sort].order as string;
+    if (order === 'asc') {
+      reqWrapper.body.sort[0][reqData.sort].order = 'desc';
+      reqWrapper.body.sort[1]['event.id'].order = 'desc';
+    } else {
+      reqWrapper.body.sort[0][reqData.sort].order = 'asc';
+      reqWrapper.body.sort[1]['event.id'].order = 'asc';
+    }
+
+    reqWrapper.body.search_after = reqData.searchBefore;
+  }
+
+  return (reqWrapper as unknown) as JsonObject;
+};
+
+export function mapToAlertResultList(
+  endpointAppContext: EndpointAppContext,
+  reqData: AlertListRequestData,
+  searchResponse: SearchResponse<AlertData>
+): AlertResultList {
+  let totalNumberOfAlerts: number = 0;
+  let totalIsLowerBound: boolean = false;
+
+  // This case is due to: https://github.com/elastic/kibana/issues/56694
+  const total: ESTotal = (searchResponse?.hits?.total as unknown) as ESTotal;
+  totalNumberOfAlerts = total?.value || 0;
+  totalIsLowerBound = total?.relation === 'gte' || false;
+
+  if (totalIsLowerBound) {
+    // This shouldn't happen, as we always try to fetch enough hits to satisfy the current request and the next page.
+    endpointAppContext.logFactory
+      .get('alerts')
+      .warn('Total hits not counted accurately. Pagination numbers may be inaccurate.');
+  }
+
+  const hits: AlertHits = searchResponse?.hits?.hits;
+  const hitLen: number = hits.length;
+
+  if (reqData.searchBefore !== undefined) {
+    // Reverse the hits if we used `search_before`.
+    hits.reverse();
```

Comment:
I'm probably misunderstanding this but do we need to reverse when using `searchBefore` or would it only be when the order is different? For example:

If the client requests page 1 with sort order ascending and it gets back the items: [1,2,3,4]
And page 2 is [5,6,7,8]
If they're on page 2 and they request page 1 using the search_before shouldn't it still return [1,2,3,4] in that order?

Reply:
So ES has no `search_before` ... we're doing it by reversing the set and using `search_after` ... so to get back to page 1 from page 2, we'll have to use `search_after 5` and reverse the sort order. That'll give us `4, 3, 2, 1`. So we have to reverse that page to get the intended results of `1, 2, 3, 4`... make sense?